code/components/citizen-devtools/src/ResourceTimeWarnings.cpp

```diff
@@ -67,7 +69,67 @@ constexpr auto tuple_slice(Cont&& t)
 	return detail::slice_impl<I1>(std::forward<Cont>(t),
 	std::make_index_sequence<I2 - I1>{});
 }
+
+static std::chrono::microseconds lastHitch;
+
+#ifdef GTA_FIVE
+#include <Hooking.h>
+#include <InputHook.h>
+
+static decltype(&SetThreadExecutionState) origSetThreadExecutionState;
+static decltype(&PeekMessageW) origPeekMessageW;
+
+static std::chrono::microseconds trackTotal;
+static bool trackIsInvalid = false;
+
+static EXECUTION_STATE WINAPI SetThreadExecutionState_Track(EXECUTION_STATE esFlags)
+{
+	if (esFlags == 3)
+	{
+		trackTotal = std::chrono::microseconds{ 0 };
+		trackIsInvalid = false;
+	}
+
+	return origSetThreadExecutionState(esFlags);
+}
+
+static BOOL WINAPI PeekMessageW_Track(LPMSG lpMsg, HWND hWnd, UINT wMsgFilterMin, UINT wMsgFilterMax, UINT wRemoveMsg)
+{
+	if (trackIsInvalid)
+	{
+		return origPeekMessageW(lpMsg, hWnd, wMsgFilterMin, wMsgFilterMax, wRemoveMsg);
+	}
+
+	auto trackStart = usec();
+	auto rv = origPeekMessageW(lpMsg, hWnd, wMsgFilterMin, wMsgFilterMax, wRemoveMsg);
```

Comment:
To prevent trapping on intentional DispatchMessage, we only account for PeekMessage time. Querying perf counter may however be slow, but granularity is of the essence here.